### ISSUE TITLE
Obtain defaults even on failed validation

### ISSUE THREAD

**Issue by rubend056** [colinhacks/zod#559]
Is there a way to obtain the defaults objects even after validation failed:
I tried:
```
const schema = z.object({
  name: z.string(),
  age: z.number().default(3),
});
const schema2 = z.object({
  name: z.string(),
  age: z.number().default(3),
  arr: z
    .array(
      z.object({
        number: z.string(),
        type: z.string().default("cell"),
      })
    )
    .nonempty()
    .default([{}] as any),
});

console.log(schema.deepPartial().default({}).parse(undefined));
// This is ok, returns { age: 3 }
console.log(schema2.deepPartial().default({}).parse(undefined));
// This is not ok, returns { age: 3 } but wanted { age: 3, arr: [{ type:'cell' }] }
```
Is there a way to actually safe parse so defaults get filled and then validate for real?

**Comment by colinhacks** [comment]
The behavior here is a little hard to explain; took me a little while to figure out why it was happening.

The takeaway is that Zod isn't respecting the default value on `arr` because the value you're providing doesn't pass validation. If you make `arr.number` optional, things work as expected, since `[{}]` is now valid:

```ts
const schema2 = z.object({
  name: z.string(),
  age: z.number().default(3),
  arr: z
    .array(
      z.object({
        number: z.string().optional(),
        type: z.string().default("cell"),
      })
    )
    .nonempty()
    .default([{}]),
});

console.log(schema2.deepPartial().default({}).parse(undefined));
//  { age: 3, arr: [{ type:'cell' }] }
```

You could also add a default value to `number` (perhaps an empty string):

```ts
const schema2 = z.object({
  name: z.string(),
  age: z.number().default(3),
  arr: z
    .array(
      z.object({
        number: z.string().default(""),
        type: z.string().default("cell"),
      })
    )
    .nonempty()
    .default([{}]),
});
console.log(schema2.deepPartial().default({}).parse(undefined));
// { age: 3, arr: [ { number: '', type: 'cell' } ] }
```

**Comment by rubend056** [comment]
Let me rephrase my question better. Is there a way to extend the deepPartial function so it makes array objects deep partial too? Or would that require too many changes and/or go against what your library is trying to do?

**Comment by colinhacks** [comment]
Just improved the `.deepPartial` implementation in zod@3.8. It now applies deeply inside array schemas as you'd expect: https://github.com/colinhacks/zod/pull/602

**Comment by tomocrafter** [comment]
The above example of
```TypeScript
const schema2 = z.object({
  name: z.string(),
  age: z.number().default(3),
  arr: z
    .array(
      z.object({
        number: z.string().default(""),
        type: z.string().default("cell"),
      })
    )
    .nonempty()
    .default([{}]),
});
console.log(schema2.deepPartial().default({}).parse(undefined));
```
seems doesnt working for me.
it always returns `{}`

**Comment by juni0r** [comment]
@tomocrafter Same here. Did you find a way to obtain defaults somehow?

**Comment by tomocrafter** [comment]
@juni0r I gave up with defining constant default value for each object. I also want there  good way to achive this...

**Comment by ifightcrime** [comment]
I'm running into this same problem, but even worse. I'm on the latest version right now (`3.19.0`) and none of the defaults are being returned:

```javascript
const schema2 = z.object({
  name: z.string(),
  age: z.number().default(3),
  arr: z
    .array(
      z.object({
        number: z.string(),
        type: z.string().default("cell"),
      })
    )
    .nonempty()
    .default([{}] as any),
});

console.log(schema2.deepPartial().default({}).parse(undefined));
// This returns an empty object for me: {}
```

So right now this means `defaults` are pretty useless to me. Am I doing something wrong?

**Comment by ifightcrime** [comment]
Interesting. It looks like #352 is close to what I'm looking for, but from the first comment on this issue it sounds like some of this should work out of the box when using `.partial()`?

> console.log(schema.deepPartial().default({}).parse(undefined));
> // This is ok, returns { age: 3 }
